gnovm/pkg/gnolang/realm.go

```diff
@@ -876,6 +883,145 @@ func (rlm *Realm) clearMarks() {
+				rlm.assertTypeIsPublic(store, head.Key.T, visited)
+			}
+			if head.Value.T != nil {
+				rlm.assertTypeIsPublic(store, head.Value.T, visited)
+			}
+		}
+	case *FuncValue:
+		if v.PkgPath != rlm.Path && isPkgPrivateFromPkgPath(store, v.PkgPath) {
+			panic("cannot persist function or method from private realm")
+		}
+		if v.Type != nil {
+			rlm.assertTypeIsPublic(store, v.Type, visited)
+		}
+		for _, capture := range v.Captures {
+			if capture.T != nil {
+				rlm.assertTypeIsPublic(store, capture.T, visited)
+			}
+		}
+	case *BoundMethodValue:
+		if v.Func.PkgPath != rlm.Path && isPkgPrivateFromPkgPath(store, v.Func.PkgPath) {
+			panic("cannot persist bound method from private realm")
+		}
+		if v.Receiver.T != nil {
+			rlm.assertTypeIsPublic(store, v.Receiver.T, visited)
+		}
+	case *Block:
+		for _, bv := range v.Values {
+			if bv.T != nil {
+				rlm.assertTypeIsPublic(store, bv.T, visited)
+			}
+		}
+		if v.Blank.T != nil {
+			rlm.assertTypeIsPublic(store, v.Blank.T, visited)
+		}
+	case *PackageValue:
+		if v.PkgPath != rlm.Path && isPkgPrivateFromPkgPath(store, v.PkgPath) {
+			panic("cannot persist package from private realm")
+		}
+	default:
+		panic(fmt.Sprintf("assertNoPrivateDeps: unhandled object type %T", v))
+	}
+}
+
+// assertTypeIsPublic ensure that the type t is not defined in a private realm.
+// it do it recursively for all types in t and have recursive guard to avoid infinite recursion on declared types.
+func (rlm *Realm) assertTypeIsPublic(store Store, t Type, visited map[TypeID]struct{}) {
+	pkgPath := ""
+
+	// NOTE: Use to avoid infinite recursion on declared types & avoid repeated checks.
+	tid := t.TypeID()
+	if _, exists := visited[tid]; exists {
+		return
+	}
+	visited[tid] = struct{}{}
+	switch tt := t.(type) {
+	case *FuncType:
+		for _, param := range tt.Params {
+			rlm.assertTypeIsPublic(store, param, visited)
+		}
+		for _, result := range tt.Results {
+			rlm.assertTypeIsPublic(store, result, visited)
+		}
+	case FieldType:
+		rlm.assertTypeIsPublic(store, tt.Type, visited)
+	case *SliceType, *ArrayType, *ChanType, *PointerType:
+		rlm.assertTypeIsPublic(store, tt.Elem(), visited)
+	case *tupleType:
+		for _, et := range tt.Elts {
+			rlm.assertTypeIsPublic(store, et, visited)
+		}
+	case *MapType:
+		rlm.assertTypeIsPublic(store, tt.Key, visited)
+		rlm.assertTypeIsPublic(store, tt.Elem(), visited)
+	case *InterfaceType:
+		pkgPath = tt.GetPkgPath()
+		for _, method := range tt.Methods {
+			rlm.assertTypeIsPublic(store, method, visited)
+		}
+	case *StructType:
+		pkgPath = tt.GetPkgPath()
+		for _, field := range tt.Fields {
+			rlm.assertTypeIsPublic(store, field, visited)
+		}
+	case *DeclaredType:
+		rlm.assertTypeIsPublic(store, tt.Base, visited)
+		for _, method := range tt.Methods {
+			rlm.assertTypeIsPublic(store, method.T, visited)
+			if mv, ok := method.V.(*FuncValue); ok {
+				rlm.assertTypeIsPublic(store, mv.Type, visited)
+			}
+		}
+		pkgPath = tt.GetPkgPath()
+	case *RefType:
```

Comment:
```suggestion
	case RefType:
```

but should it happen here? RefType is filled while restoring objects.

Reply:
i think it's good to keep it there but add a panic, "should not happen, xxx" is good.

Reply:
[e4826b3](https://github.com/gnolang/gno/pull/4422/commits/e4826b3124b8ac7fa4571790f91eea5ab593aa24)
(i think it does not matter because default case is a panic with type printed)